ray-curator/ray_curator/stages/image/deduplication/removal.py

```diff
@@ -0,0 +1,104 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from dataclasses import dataclass, field
+from typing import Any
+
+import pyarrow as pa
+import pyarrow.dataset as ds
+from loguru import logger
+
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.tasks import ImageBatch
+
+
+@dataclass
+class ImageDuplicatesRemovalStage(ProcessingStage[ImageBatch, ImageBatch]):
+    """Filter stage that removes images whose IDs appear in a Parquet file.
+
+    The Parquet file must contain a column with image identifiers; by default this
+    column is assumed to be ``id`` to match writer metadata. You can change
+    the column name via ``duplicate_id_field``.
+
+    Args:
+        removal_parquets_dir: Directory containing Parquet files with image IDs to remove
+        duplicate_id_field: Name of the column containing image IDs to remove
+        verbose: Whether to log verbose output
+        num_workers_per_node: Number of workers per node for the stage. This is sometimes needed
+            to avoid OOM when concurrently running actors on one node loading the same removal
+            parquet files into memory.
+    """
+
+    removal_parquets_dir: str
+    duplicate_id_field: str = "id"
+    verbose: bool = False
+    num_workers_per_node: int | None = None
+
+    _name: str = "image_dedup_filter"
+
+    # Internal cache
+    _ids_to_remove: set[str] = field(default_factory=set)
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        return ["data"], []
+
+    def setup(self, _worker_metadata=None) -> None:  # noqa: ANN001
+        removal_parquets = [os.path.join(self.removal_parquets_dir, f) for f in os.listdir(self.removal_parquets_dir) if f.endswith(".parquet")]
+        if not removal_parquets:
+            msg = f"No parquet files found in {self.removal_parquets_dir}"
+            logger.error(msg)
+            raise FileNotFoundError(msg)
+
+        # Read entire directory using a Dataset with explicit schema for the ID column.
+        # This avoids Arrow attempting to unify string and null types from empty files.
+        schema = pa.schema([pa.field(self.duplicate_id_field, pa.string())])
+        dataset = ds.dataset(self.removal_parquets_dir, format="parquet", schema=schema)
+        table = dataset.to_table(columns=[self.duplicate_id_field])
+        ids_array = table[self.duplicate_id_field].to_pylist()
+        self._ids_to_remove.update(ids_array)
```

Comment:
I think we should  read a single column here (do not read extra fields) and do it via pandas for simplicity. Speed is the same as they both use the same engine.   


```python3
df = pd.read_parquet(
    self.removal_parquets_dir,
    columns=[self.duplicate_id_field],
    engine="pyarrow",
)

ids_array = df[self.duplicate_id_field].astype("string").to_list()  
```

CC: @praateekmahajan  for input for removal stage inputs.